### PR TITLE
Custom join on conditions

### PR DESCRIPTION
**Pull request by billy1624**
## PR Info

- Closes https://github.com/SeaQL/sea-orm/issues/160

## Adds

- [x] `RelationDef` and `RelationBuilder` with `on_condition` field

### CODE REVIEW

File: src/query/join.rs [hunk 1]
```diff
@@ -355,4 +366,62 @@ mod tests {
...
+                r#"SELECT `cake`.`id` AS `A_id`, `cake`.`name` AS `A_name`,"#,
+                r#"`r2`.`id` AS `B_id`, `r2`.`name` AS `B_name`"#,
+                r#"FROM `cake`"#,
+                r#"LEFT JOIN `cake_filling` AS `r0` ON `cake`.`id` = `r0`.`cake_id`"#,
+                r#"LEFT JOIN `filling` AS `r1` ON `r0`.`filling_id` = `r1`.`id`"#,
+                r#"LEFT JOIN `vendor` AS `r2` ON `r1`.`vendor_id` = `r2`.`id`"#,
+            ]
+            .join(" ")
+        );
+    }
```
Comment (billy1624) [inline src/query/join.rs:426]: I haven't figured out how should we model this. As discussed on
- https://github.com/SeaQL/sea-orm/issues/160#issuecomment-1139522611

CC @tyt2y3

**Comment by liberwang1013** [comment]
thank you for inviting me , this implement is great. I have one more question about how it handle multiple join conditions? Can you  gave me more examples?

File: src/entity/relation.rs [hunk 2]
```diff
@@ -42,7 +42,7 @@ where
 }
 
 /// Defines a relationship
-#[derive(Debug)]
+#[allow(missing_debug_implementations)]
```
Comment (tyt2y3) [inline src/entity/relation.rs:45]: Actually, why not impl Debug manually?

File: src/tests_cfg/entity_linked.rs [hunk 3]
```diff
@@ -32,3 +33,27 @@ impl Linked for CakeToFillingVendor {
...
+                    Expr::tbl(left, super::cake::Column::Name)
+                        .like("%cheese%")
+                        .into_condition()
+                })
+                .rev(),
+            super::cake_filling::Relation::Filling.def(),
+            super::filling::Relation::Vendor.def(),
+        ]
+    }
+}
```
Comment (tyt2y3) [inline src/tests_cfg/entity_linked.rs:59]: I think you nailed that in terms of API ergonomics. And yeah no reason we cannot bake that into RelationDef. The only concern I had was it's kind of complicated to reason about.

And I'd appreciate that it to be documented somewhere what the second parameter is, or at least `left, _right`

Comment (tyt2y3) [reply src/tests_cfg/entity_linked.rs:59]: Actually, does that still work when a `rev` is applied?

Comment (billy1624) [reply src/tests_cfg/entity_linked.rs]: With `rev` or without `rev`. The behaviour is the same. 
- `left` represent the `from_tbl` inside `RelationDef`
- `right` represent the `to_tbl` inside `RelationDef`

Examples: https://github.com/SeaQL/sea-orm/pull/793/files#diff-be3eeec8b487c64c2f782fc15091165849d2677415f718a648dfcc1221044152R385-R426

**Comment by billy1624** [comment]
> thank you for inviting me , this implement is great. I have one more question about how it handle multiple join conditions? Can you gave me more examples?

Hey @liberwang1013, for example, https://github.com/SeaQL/sea-orm/pull/793/commits/3efe1611a2bb0b658aab78ffc9e22005760cbd96?

File: src/entity/relation.rs [hunk 4]
```diff
@@ -97,9 +155,19 @@ impl RelationDef {
...
             on_update: self.on_update,
+            on_condition: self.on_condition,
             fk_name: None,
         }
     }
+
+    /// Set custom join ON condition
+    pub fn on_condition<F>(mut self, f: F) -> Self
+    where
+        F: Fn(DynIden, DynIden) -> Condition + 'static,
```
Comment (tyt2y3) [inline src/entity/relation.rs:166]: May be typedef this signature and add some documents?

**Review by tyt2y3 — APPROVED** [review]

**Review by tyt2y3 — CHANGES_REQUESTED** [review]
API Docs

File: src/entity/relation.rs (hunk 4 above)
Comment (tyt2y3) [reply src/entity/relation.rs:166]: Ah trait alias is still nightly.

Comment (tyt2y3) [reply src/entity/relation.rs:166]: But we need to at least document what is fn is supposed to do and what the parameters mean

File: src/entity/relation.rs [hunk 5]
```diff
@@ -97,9 +155,19 @@ impl RelationDef {
             is_owner: !self.is_owner,
             on_delete: self.on_delete,
             on_update: self.on_update,
+            on_condition: self.on_condition,
             fk_name: None,
         }
     }
+
+    /// Set custom join ON condition
```
Comment (tyt2y3) [inline src/entity/relation.rs:163]: Doc test example needed

File: src/entity/relation.rs [hunk 6]
```diff
@@ -167,6 +237,15 @@ where
         self
     }
 
+    /// Set custom join ON condition
```
Comment (tyt2y3) [inline src/entity/relation.rs:240]: Docs

**Comment by billy1624** [comment]
Docs added

**Review by tyt2y3 — APPROVED** [review]

**Comment by tyt2y3** [comment]
Great work!

**Pull request merged**